Eval bug: example/finetune.cpp crashing

### Name and Version

./build/bin/llama-finetune --version
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 2 CUDA devices:
  Device 0: NVIDIA H100 NVL, compute capability 9.0, VMM: yes
  Device 1: NVIDIA H100 NVL, compute capability 9.0, VMM: yes
version: 5754 (2bf9d539)
built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu

### Operating systems

Linux

### GGML backends

CPU

### Hardware

some_user@some_hardware:~$ lscpu

Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          52 bits physical, 57 bits virtual
  Byte Order:             Little Endian
CPU(s):                   256
  On-line CPU(s) list:    0-255
Vendor ID:                AuthenticAMD
  Model name:             AMD EPYC 9554 64-Core Processor
    CPU family:           25
    Model:                17
    Thread(s) per core:   2
    Core(s) per socket:   64
    Socket(s):            2
    Stepping:             1
    Frequency boost:      enabled
    CPU max MHz:          3100.0000
    CPU min MHz:          1500.0000
    BogoMIPS:             6199.89
    Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_
                          opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 f
                          ma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a
                           misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l
                          3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_
                          a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsav
                          es cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat
                          npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif
                           v_spec_ctrl avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdp
                          id overflow_recov succor smca fsrm flush_l1d
Virtualization features:
  Virtualization:         AMD-V
Caches (sum of all):
  L1d:                    4 MiB (128 instances)
  L1i:                    4 MiB (128 instances)
  L2:                     128 MiB (128 instances)
  L3:                     512 MiB (16 instances)
NUMA:
  NUMA node(s):           2
  NUMA node0 CPU(s):      0-63,128-191
  NUMA node1 CPU(s):      64-127,192-255
Vulnerabilities:
  Gather data sampling:   Not affected
  Itlb multihit:          Not affected
  L1tf:                   Not affected
  Mds:                    Not affected
  Meltdown:               Not affected
  Mmio stale data:        Not affected
  Reg file data sampling: Not affected
  Retbleed:               Not affected
  Spec rstack overflow:   Mitigation; safe RET
  Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
  Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not aff
                          ected
  Srbds:                  Not affected
  Tsx async abort:        Not affected

### Models

[Dolphin3.0-Llama3.2-1B-f32.gguf](https://huggingface.co/bartowski/Dolphin3.0-Llama3.2-1B-GGUF)
[smollm2-135M-base-f32.gguf](https://huggingface.co/mav23/SmolLM2-135M-Instruct-GGUF)

I was really hoping this model would work, seeing how it's the one being used in `examples/training/README.md`
[llama3.2-1b-f32.gguf](https://huggingface.co/bartowski/Llama-3.2-1B-Instruct-GGUF)

### Problem description & steps to reproduce

[BUILD]:
From the file: docs/build.md:21-22
cmake -B build
cmake --build build --config Release

[RUN]:
From the file: examples/training/README.md:12-14
export model_name=llama_3.2-1b && export quantization=f32
./build/bin/llama-finetune --file datasets/wikitext-2-raw/wiki.test.raw -ngl 999 --model models/${model_name}-${quantization}.gguf -c 512 -b 512 -ub 512

I have tried using these models:
Dolphin3.0-Llama3.2-1B-f32.gguf
llama3.2-1b-f32.gguf
smollm2-135M-base-f32.gguf

Has anyone else encountered this issue when trying to run the finetune example as written?

I am trying to run the finetune example as-is following the information given in the docs/build.md and README.md and examples/training/README.md

### First Bad Commit

Unknown.

### Relevant log output

```shell
[RESULT]:
some_user@some_hardware:~/dev/some_project/llama.cpp$ ./build/bin/llama-finetune --file datasets/wikitext-2-raw/wiki.test.raw -ngl 999 --model models/${model_name}-${quantization}.gguf -c 512 -b 512 -ub 512
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 2 CUDA devices:
  Device 0: NVIDIA H100 NVL, compute capability 9.0, VMM: yes
  Device 1: NVIDIA H100 NVL, compute capability 9.0, VMM: yes
main: force disabling memory mapping because it would result in-read-only pointers to the weights
main: force changing k cache type to f32 due to a lack of f16 support for OUT_PROD
main: force changing v cache type to f32 due to a lack of f16 support for OUT_PROD
build: 5754 (2bf9d539) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
llama_model_load_from_file_impl: using device CUDA0 (NVIDIA H100 NVL) - 94805 MiB free
llama_model_load_from_file_impl: using device CUDA1 (NVIDIA H100 NVL) - 94805 MiB free
llama_model_loader: loaded meta data with 33 key-value pairs and 147 tensors from models/llama3.2-1b-f32.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Llama 3.2 1b
llama_model_loader: - kv   3:                       general.organization str              = Meta Llama
llama_model_loader: - kv   4:                           general.basename str              = llama-3.2
llama_model_loader: - kv   5:                         general.size_label str              = 1B
llama_model_loader: - kv   6:                            general.license str              = llama3.2
llama_model_loader: - kv   7:                   general.base_model.count u32              = 1
llama_model_loader: - kv   8:                  general.base_model.0.name str              = Llama 3.2 1B
llama_model_loader: - kv   9:          general.base_model.0.organization str              = Meta Llama
llama_model_loader: - kv  10:              general.base_model.0.repo_url str              = https://huggingface.co/meta-llama/Lla...
llama_model_loader: - kv  11:                               general.tags arr[str,2]       = ["gguf-connector", "text-generation"]
llama_model_loader: - kv  12:                          llama.block_count u32              = 16
llama_model_loader: - kv  13:                       llama.context_length u32              = 131072
llama_model_loader: - kv  14:                     llama.embedding_length u32              = 2048
llama_model_loader: - kv  15:                  llama.feed_forward_length u32              = 8192
llama_model_loader: - kv  16:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv  17:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv  18:                       llama.rope.freq_base f32              = 500000.000000
llama_model_loader: - kv  19:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  20:                 llama.attention.key_length u32              = 64
llama_model_loader: - kv  21:               llama.attention.value_length u32              = 64
llama_model_loader: - kv  22:                          general.file_type u32              = 0
llama_model_loader: - kv  23:                           llama.vocab_size u32              = 128256
llama_model_loader: - kv  24:                 llama.rope.dimension_count u32              = 64
llama_model_loader: - kv  25:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  26:                         tokenizer.ggml.pre str              = llama-bpe
llama_model_loader: - kv  27:                      tokenizer.ggml.tokens arr[str,128256]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  28:                  tokenizer.ggml.token_type arr[i32,128256]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  29:                      tokenizer.ggml.merges arr[str,280147]  = ["Ġ Ġ", "Ġ ĠĠĠ", "ĠĠ ĠĠ", "...
llama_model_loader: - kv  30:                tokenizer.ggml.bos_token_id u32              = 128000
llama_model_loader: - kv  31:                tokenizer.ggml.eos_token_id u32              = 128001
llama_model_loader: - kv  32:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:  147 tensors
print_info: file format = GGUF V3 (latest)
print_info: file type   = all F32
print_info: file size   = 4.60 GiB (32.00 BPW)
load: special tokens cache size = 256
load: token to piece cache size = 0.7999 MB
print_info: arch             = llama
print_info: vocab_only       = 0
print_info: n_ctx_train      = 131072
print_info: n_embd           = 2048
print_info: n_layer          = 16
print_info: n_head           = 32
print_info: n_head_kv        = 8
print_info: n_rot            = 64
print_info: n_swa            = 0
print_info: is_swa_any       = 0
print_info: n_embd_head_k    = 64
print_info: n_embd_head_v    = 64
print_info: n_gqa            = 4
print_info: n_embd_k_gqa     = 512
print_info: n_embd_v_gqa     = 512
print_info: f_norm_eps       = 0.0e+00
print_info: f_norm_rms_eps   = 1.0e-05
print_info: f_clamp_kqv      = 0.0e+00
print_info: f_max_alibi_bias = 0.0e+00
print_info: f_logit_scale    = 0.0e+00
print_info: f_attn_scale     = 0.0e+00
print_info: n_ff             = 8192
print_info: n_expert         = 0
print_info: n_expert_used    = 0
print_info: causal attn      = 1
print_info: pooling type     = 0
print_info: rope type        = 0
print_info: rope scaling     = linear
print_info: freq_base_train  = 500000.0
print_info: freq_scale_train = 1
print_info: n_ctx_orig_yarn  = 131072
print_info: rope_finetuned   = unknown
print_info: ssm_d_conv       = 0
print_info: ssm_d_inner      = 0
print_info: ssm_d_state      = 0
print_info: ssm_dt_rank      = 0
print_info: ssm_dt_b_c_rms   = 0
print_info: model type       = 1B
print_info: model params     = 1.24 B
print_info: general.name     = Llama 3.2 1b
print_info: vocab type       = BPE
print_info: n_vocab          = 128256
print_info: n_merges         = 280147
print_info: BOS token        = 128000 '<|begin_of_text|>'
print_info: EOS token        = 128001 '<|end_of_text|>'
print_info: EOT token        = 128009 '<|eot_id|>'
print_info: EOM token        = 128008 '<|eom_id|>'
print_info: LF token         = 198 'Ċ'
print_info: EOG token        = 128001 '<|end_of_text|>'
print_info: EOG token        = 128008 '<|eom_id|>'
print_info: EOG token        = 128009 '<|eot_id|>'
print_info: max token length = 256
load_tensors: loading model tensors, this can take a while... (mmap = false)
load_tensors: offloading 16 repeating layers to GPU
load_tensors: offloading output layer to GPU
load_tensors: offloaded 17/17 layers to GPU
load_tensors:    CUDA_Host model buffer size =  1002.00 MiB
load_tensors:        CUDA0 model buffer size =  2088.14 MiB
load_tensors:        CUDA1 model buffer size =  2626.12 MiB
..................................................................
llama_context: constructing llama_context
llama_context: n_seq_max     = 1
llama_context: n_ctx         = 512
llama_context: n_ctx_per_seq = 512
llama_context: n_batch       = 512
llama_context: n_ubatch      = 512
llama_context: causal_attn   = 1
llama_context: flash_attn    = 0
llama_context: freq_base     = 500000.0
llama_context: freq_scale    = 1
llama_context: n_ctx_per_seq (512) < n_ctx_train (131072) -- the full capacity of the model will not be utilized
llama_context:  CUDA_Host  output buffer size =     0.49 MiB
llama_kv_cache_unified:      CUDA0 KV buffer size =    18.00 MiB
llama_kv_cache_unified:      CUDA1 KV buffer size =    14.00 MiB
llama_kv_cache_unified: size =   32.00 MiB (   512 cells,  16 layers,  1 seqs), K (f32):   16.00 MiB, V (f32):   16.00 MiB
llama_context: pipeline parallelism enabled (n_copies=4)
llama_context:      CUDA0 compute buffer size =    68.01 MiB
llama_context:      CUDA1 compute buffer size =   290.52 MiB
llama_context:  CUDA_Host compute buffer size =     8.02 MiB
llama_context: graph nodes  = 582
llama_context: graph splits = 3
common_init_from_params: setting dry_penalty_last_n to ctx_size = 512
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)

system_info: n_threads = 128 (n_threads_batch = 128) / 256 | CUDA : ARCHS = 500,610,700,750,800,860,890 | USE_GRAPHS = 1 | PEER_MAX_BATCH_SIZE = 128 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | AVX512 = 1 | AVX512_VBMI = 1 | AVX512_VNNI = 1 | AVX512_BF16 = 1 | LLAMAFILE = 1 | OPENMP = 1 | REPACK = 1 |
/home/some_user/dev/some_project/llama.cpp/ggml/src/ggml-backend.cpp:750: pre-allocated tensor (AdamW step for output_norm.weight) in a buffer (CUDA1) that cannot run the operation (OPT_STEP_ADAMW)
[New LWP 134852]
[New LWP 134859]
[New LWP 134860]
[New LWP 134861]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f72f49e742f in __GI___wait4 (pid=134862, stat_loc=0x0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30      ../sysdeps/unix/sysv/linux/wait4.c: No such file or directory.
#0  0x00007f72f49e742f in __GI___wait4 (pid=134862, stat_loc=0x0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30      in ../sysdeps/unix/sysv/linux/wait4.c
#1  0x00007f72f4f2707b in ggml_print_backtrace () from /home/some_user/dev/some_project/llama.cpp/build/bin/libggml-base.so
#2  0x00007f72f4f2718e in ggml_abort () from /home/some_user/dev/some_project/llama.cpp/build/bin/libggml-base.so
#3  0x00007f72f4f3a394 in ggml_backend_sched_backend_id_from_cur(ggml_backend_sched*, ggml_tensor*) () from /home/some_user/dev/some_project/llama.cpp/build/bin/libggml-base.so
#4  0x00007f72f4f3badf in ggml_backend_sched_split_graph(ggml_backend_sched*, ggml_cgraph*) [clone .part.0] () from /home/some_user/dev/some_project/llama.cpp/build/bin/libggml-base.so
#5  0x00007f72f4f3d1fb in ggml_backend_sched_alloc_graph () from /home/some_user/dev/some_project/llama.cpp/build/bin/libggml-base.so
#6  0x00007f72f4f4116c in ggml_opt_alloc () from /home/some_user/dev/some_project/llama.cpp/build/bin/libggml-base.so
#7  0x00007f72f507ac87 in llama_context::opt_epoch_iter(ggml_opt_dataset*, ggml_opt_result*, std::vector<int, std::allocator<int> > const&, std::vector<int, std::allocator<int> > const&, llama_batch&, void (*)(bool, ggml_opt_context*, ggml_opt_dataset*, ggml_opt_result*, long, long, long), bool, long, long, long) () from /home/some_user/dev/some_project/llama.cpp/build/bin/libllama.so
#8  0x00007f72f507b23d in llama_context::opt_epoch(ggml_opt_dataset*, ggml_opt_result*, ggml_opt_result*, long, void (*)(bool, ggml_opt_context*, ggml_opt_dataset*, ggml_opt_result*, long, long, long), void (*)(bool, ggml_opt_context*, ggml_opt_dataset*, ggml_opt_result*, long, long, long)) () from /home/some_user/dev/some_project/llama.cpp/build/bin/libllama.so
#9  0x00005606a28d1757 in main ()
[Inferior 1 (process 134851) detached]
Aborted (core dumped)
```

FWIW, my attempt with the latest llama.cpp 
time ./build/bin/llama-finetune --model ~/SmolLM2-135/SmolLM2-135-f32.gguf --file ~/Downloads/wiki.test.raw -ngl 
0 -b 512 -c 512 -ub 512 --device none                                                                                                       
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no                                                                                                  ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no                                                                                                  
ggml_cuda_init: found 1 CUDA devices:                                                                                                         Device 0: NVIDIA GeForce RTX 4090, compute capability 8.9, VMM: yes                                                                       
register_backend: registered backend CUDA (1 devices)                                                                                       
register_device: registered device CUDA0 (NVIDIA GeForce RTX 4090)                                                                          
register_backend: registered backend BLAS (1 devices)                                                                                       
register_device: registered device BLAS (MKL)                                                                                               
register_backend: registered backend CPU (1 devices)                                                                                        
register_device: registered device CPU (AMD EPYC 7R32 48-Core Processor)                     
load_backend: failed to find ggml_backend_init in /home/b/llama.cpp/build/bin/libggml-blas.so
load_backend: failed to find ggml_backend_init in /home/b/llama.cpp/build/bin/libggml-cuda.so    
load_backend: failed to find ggml_backend_init in /home/b/llama.cpp/build/bin/libggml-cpu.so     
main: force disabling memory mapping because it would result in-read-only pointers to the weights
main: force changing k cache type to f32 due to a lack of f16 support for OUT_PROD
main: force changing v cache type to f32 due to a lack of f16 support for OUT_PROD 
build: 5797 (de569441) with Intel(R) oneAPI DPC++/C++ Compiler 2025.1.1 (2025.1.1.20250418) for x86_64-unknown-linux-gnu (debug)
llama_model_loader: loaded meta data with 29 key-value pairs and 272 tensors from /home/b/SmolLM2-135/SmolLM2-135-f32.gguf (version GGUF V3 
(latest))                                                                                                                                   
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.type str              = model                                           
llama_model_loader: - kv   2:                         general.size_label str              = 135M                                            
llama_model_loader: - kv   3:                            general.license str              = apache-2.0                                      
llama_model_loader: - kv   4:                          general.languages arr[str,1]       = ["en"]                                          
llama_model_loader: - kv   5:                          llama.block_count u32              = 30                                              llama_model_loader: - kv   6:                       llama.context_length u32              = 8192                                            
llama_model_loader: - kv   7:                     llama.embedding_length u32              = 576                                             
llama_model_loader: - kv   8:                  llama.feed_forward_length u32              = 1536                                            
llama_model_loader: - kv   9:                 llama.attention.head_count u32              = 9                                               
llama_model_loader: - kv  10:              llama.attention.head_count_kv u32              = 3                                               
llama_model_loader: - kv  11:                       llama.rope.freq_base f32              = 100000.000000                                   
llama_model_loader: - kv  12:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010                                        
llama_model_loader: - kv  13:                 llama.attention.key_length u32              = 64                                              
llama_model_loader: - kv  14:               llama.attention.value_length u32              = 64                                              
llama_model_loader: - kv  15:                          general.file_type u32              = 0                                               
llama_model_loader: - kv  16:                           llama.vocab_size u32              = 49152                                           llama_model_loader: - kv  17:                 llama.rope.dimension_count u32              = 64                                              
llama_model_loader: - kv  18:               general.quantization_version u32              = 2                                               llama_model_loader: - kv  19:                       tokenizer.ggml.model str              = gpt2                                            
llama_model_loader: - kv  20:                         tokenizer.ggml.pre str              = smollm                                          
llama_model_loader: - kv  21:                      tokenizer.ggml.tokens arr[str,49152]   = ["<|endoftext|>", "<|im_start|>", "<|...        
llama_model_loader: - kv  22:                  tokenizer.ggml.token_type arr[i32,49152]   = [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, ...        
llama_model_loader: - kv  23:                      tokenizer.ggml.merges arr[str,48900]   = ["Ġ t", "Ġ a", "i n", "h e", "Ġ Ġ...            
llama_model_loader: - kv  24:                tokenizer.ggml.bos_token_id u32              = 0                                               
llama_model_loader: - kv  25:                tokenizer.ggml.eos_token_id u32              = 0
llama_model_loader: - kv  26:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  27:            tokenizer.ggml.add_space_prefix bool             = false
llama_model_loader: - kv  28:               tokenizer.ggml.add_bos_token bool             = false
llama_model_loader: - type  f32:  272 tensors                                                                                               
print_info: file format = GGUF V3 (latest)                                                                                                  
print_info: file type   = all F32     
(...skipping...)
lama_context: freq_scale    = 1                                                                                                    [0/1529]
llama_context: n_ctx_per_seq (512) < n_ctx_train (8192) -- the full capacity of the model will not be utilized                              
llama_context:        CPU  output buffer size =     0.19 MiB                                                                                
llama_kv_cache_unified:        CPU KV buffer size =    22.50 MiB                                                                            
llama_kv_cache_unified: size =   22.50 MiB (   512 cells,  30 layers,  1 seqs), K (f32):   11.25 MiB, V (f32):   11.25 MiB                  
llama_context:        CPU compute buffer size =   101.51 MiB                                                                                
llama_context: graph nodes  = 1056                                                                                                          
llama_context: graph splits = 422 (with bs=512), 1 (with bs=1)                                                                              
common_init_from_params: setting dry_penalty_last_n to ctx_size = 512                                                                       
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)                                  
                                                                                                                                            
system_info: n_threads = 180 (n_threads_batch = 180) / 180 | CUDA : ARCHS = 890 | USE_GRAPHS = 1 | PEER_MAX_BATCH_SIZE = 128 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | REPACK = 1 |                              
output_reserve: reallocating output buffer from size 0.19 MiB to 96.00 MiB                                                                  
ggml_compute_backward: unsupported ggml op for backward pass: GLU                                                                           
/home/b/llama.cpp/ggml/src/ggml.c:6018: fatal error                                                                                         
/home/b/llama.cpp/build/bin/libggml-base.so(+0x12528) [0x7fa8d2573528]                                                                      
/home/b/llama.cpp/build/bin/libggml-base.so(ggml_print_backtrace+0x231) [0x7fa8d2573501]                                                    
/home/b/llama.cpp/build/bin/libggml-base.so(ggml_abort+0xd0) [0x7fa8d25725f0]                                                               
/home/b/llama.cpp/build/bin/libggml-base.so(ggml_build_backward_expand+0x2597) [0x7fa8d2582b07]                                             
/home/b/llama.cpp/build/bin/libggml-base.so(+0x2f91c) [0x7fa8d259091c]                                                                      
/home/b/llama.cpp/build/bin/libggml-base.so(ggml_opt_alloc+0xc1) [0x7fa8d2591831]                                                           
/home/b/llama.cpp/build/bin/libllama.so(_ZN13llama_context14opt_epoch_iterEP16ggml_opt_datasetP15ggml_opt_resultRKSt6vectorIiSaIiEES8_R11llama_batchPFvbP16ggml_opt_contextS1_S3_lllEblll+0x394) [0x7fa8d3aefb34]                                                                       
/home/b/llama.cpp/build/bin/libllama.so(_ZN13llama_context9opt_epochEP16ggml_opt_datasetP15ggml_opt_resultS3_lPFvbP16ggml_opt_contextS1_S3_lllES7_+0x20c) [0x7fa8d3af001c]                                                                                                              
./build/bin/llama-finetune() [0x41a451]                                                                                                     
/lib/x86_64-linux-gnu/libc.so.6(+0x29ca8) [0x7fa8cd943ca8]                                                                                  
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x85) [0x7fa8cd943d65]                                                                    
./build/bin/llama-finetune() [0x41a0a1]                                                                                                     
Aborted

--- a/ggml/src/ggml.c
+++ b/ggml/src/ggml.c
@@ -6009,13 +6009,28 @@ static void ggml_compute_backward(
             }
             GGML_ASSERT(!src1_needs_grads && "backward pass for labels not implemented");
         } break;
+        case GGML_OP_GLU: {
+            switch (ggml_get_glu_op(tensor)) {
+                case GGML_GLU_OP_SWIGLU: {
+                    if (src0_needs_grads) {
+                        GGML_ASSERT(src1 && "backward pass only implemented for split swiglu");
+                        ggml_add_or_set(ctx, cgraph, isrc0, ggml_silu_back(ctx, ggml_mul(ctx, grad, src1), src0));
+                    }
+                    if (src1_needs_grads) {
+                        ggml_add_or_set(ctx, cgraph, isrc1, ggml_mul(ctx, ggml_silu(ctx, src0), grad));
+                    }
+                } break;
+                default: {
+                    GGML_ABORT("unsupported glu op for backward pass: %s", ggml_glu_op_name(ggml_get_glu_op(tensor)));
+                } //break;
+            }
+        } break;
         case GGML_OP_NONE: {
             // noop
         } break;
         case GGML_OP_COUNT:
         default: {
-            fprintf(stderr, "%s: unsupported ggml op for backward pass: %s\n", __func__, ggml_op_name(tensor->op));
-            GGML_ABORT("fatal error");
+            GGML_ABORT("%s: unsupported ggml op for backward pass: %s\n", __func__, ggml_op_name(tensor->op));
         } //break;
     }
 

--- a/tests/test-backend-ops.cpp
+++ b/tests/test-backend-ops.cpp
@@ -1175,21 +1175,25 @@ struct test_glu_split : public test_case {
         if (v & 1) {
             auto ne = ne_a; ne[0] *= 3;
             a = ggml_new_tensor(ctx, type, 4, ne.data());
+            ggml_set_param(a);
             ggml_set_name(a, "a");
 
             a = ggml_view_4d(ctx, a, ne_a[0], ne_a[1], ne_a[2], ne_a[3], a->nb[1], a->nb[2], a->nb[3], 0);
             ggml_set_name(a, "view_of_a");
 
             b = ggml_new_tensor(ctx, type, 4, ne.data());
+            ggml_set_param(b);
             ggml_set_name(b, "b");
 
             b = ggml_view_4d(ctx, b, ne_a[0], ne_a[1], ne_a[2], ne_a[3], b->nb[1], b->nb[2], b->nb[3], 0);
             ggml_set_name(a, "view_of_b");
         } else {
             a = ggml_new_tensor(ctx, type, 4, ne_a.data());
+            ggml_set_param(a);
             ggml_set_name(a, "a");
 
             b = ggml_new_tensor(ctx, type, 4, ne_a.data());
+            ggml_set_param(b);
             ggml_set_name(b, "b");
         }
 

